File: sem 4/JIPP/lab3/zad2/DHSetup.java
package zad2;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class DHSetup<T extends ModuloField> {

    private T generator;

    public DHSetup() {
        long characteristic = ModuloField.getCharacteristic();
        Random rd = new Random();
        long generatorValue;

        List<Long> divisors = new ArrayList<>();
        for (long i = 2; i * i < characteristic - 1; i++) {
            if ((characteristic - 1) % i == 0) {
                divisors.add(i);
            }
        }

        int s = divisors.size();
        for (int i = 0; i < s; i++) {
            divisors.add((characteristic - 1) / divisors.get(i));
        }

        List<Long> primeDivisors = new ArrayList<>();
        for (int i = 0; i < divisors.size(); i++) {
            boolean isPrime = true;
            for (int j = 2; j * j <= divisors.get(i); j++) {
                if (divisors.get(i) % j == 0) {
                    isPrime = false;
                    break;
                }
            }
            if (isPrime) {
                primeDivisors.add(divisors.get(i));
            }
        }

        do {
            generatorValue = rd.nextLong(characteristic - 1) + 1;
            generator = (T) new ModuloField(generatorValue);
        } while (!generatorOk(primeDivisors));
    }

    private boolean generatorOk(List<Long> primeDivisors) {
        for (long primeDivisor : primeDivisors) {
            if (power(generator, (T.getCharacteristic() - 1) / primeDivisor).equals((T) new ModuloField(1))) {
                return false;
            }
        }
        return true;
    }

    public T getGenerator() {
        return generator;
    }

    public T power(T a, long b) {
        T result = (T) new ModuloField(1);
        while (b > 0) {
            if (b % 2 == 1) {
                result = (T) ModuloField.multiply(result, a);
            }
            a = (T) ModuloField.multiply(a, a);
            b /= 2;
        }
        return result;
    }
}

File: sem 4/JIPP/lab3/zad3/DHSetup.py
from typing import TypeVar, Generic
from random import randint
from ModuloField import ModuloField

T = TypeVar('T')

class DHSetup(Generic[T]):
    def __init__(self) -> None:
        self.generator = None
        divisors = [i for i in range(2, ModuloField.characteristic - 1) if (ModuloField.characteristic - 1) % i == 0]
        prime_divisors = [d for d in divisors if all(d % j != 0 for j in range(2, int(d**0.5) + 1))]
        while True:
            self.generator = ModuloField(randint(1, ModuloField.characteristic - 1))
            if self.generator_ok(prime_divisors):
                break

    def generator_ok(self, prime_divisors: list[int]) -> bool:
        for prime_divisor in prime_divisors:
            if self.power(self.generator, (ModuloField.characteristic - 1) // prime_divisor) == ModuloField(1):
                return False
        return True

    def get_generator(self) -> T:
        return self.generator

    @staticmethod
    def power(a: ModuloField, b: int) -> ModuloField:
        result = ModuloField(1)
        while b > 0:
            if b % 2 == 1:
                result = result * a
            a = a * a
            b //= 2
        return result